python/preemo/worker/_artifact_manager.py

```diff
@@ -134,37 +227,81 @@ def get_artifacts(self, artifact_ids: List[ArtifactId]) -> List[bytes]:
+            BatchGetArtifactPartRequest(
+                configs_by_artifact_id=configs_by_artifact_id_value
+            )
         )
 
-        results: List[bytes] = []
-        for (
-            artifact_id,
-            result,
-        ) in get_artifact_part_response.results_by_artifact_id.items():
-            config = configs_by_artifact_id[artifact_id]
-            ensure_keys_match(
-                expected=config.metadatas_by_part_number,
-                actual=result.metadatas_by_part_number,
+        with concurrent.futures.ThreadPoolExecutor(
+            max_workers=self._max_download_threads
+        ) as executor:
+            futures_by_artifact_id_and_part_number: Dict[
+                ArtifactId, Dict[int, concurrent.futures.Future]
+            ] = {}
+            for (
+                artifact_id_value,
+                artifact_part_result,
+            ) in get_artifact_part_response.results_by_artifact_id.items():
+                config = configs_by_artifact_id_value[artifact_id_value]
+                ensure_keys_match(
+                    expected=config.metadatas_by_part_number,
+                    actual=artifact_part_result.metadatas_by_part_number,
+                )
+
+                futures_by_part_number: Dict[int, concurrent.futures.Future] = {}
+                for (
+                    part_number,
+                    metadata,
+                ) in artifact_part_result.metadatas_by_part_number.items():
+                    futures_by_part_number[part_number] = executor.submit(
+                        ArtifactManager._download_content,
+                        url=metadata.download_signed_url,
+                    )
+
+                futures_by_artifact_id_and_part_number[
+                    ArtifactId(value=artifact_id_value)
+                ] = futures_by_part_number
+
+            futures = [
+                future
+                for futures_by_part_number in futures_by_artifact_id_and_part_number.values()
+                for future in futures_by_part_number.values()
+            ]
+            done, not_done = concurrent.futures.wait(
+                futures,
+                return_when=concurrent.futures.ALL_COMPLETED,
             )
 
-            # TODO(adrian@preemo.io, 03/20/2023): this hard-coded part number will need to change for multi-part upload
-            # metadata = result.metadatas_by_part_number[1]
+            if len(not_done) != 0:
+                raise Exception("expected incomplete future set to be empty")
+
+            if len(done) != len(futures):
+                raise Exception("expected all futures to have completed")
+
+            results: List[bytes] = []
+            for artifact_id in artifact_ids:
+                futures_by_part_number = futures_by_artifact_id_and_part_number[
+                    artifact_id
+                ]
+
+                result = bytearray()
```

Comment:
yeah I wanna leave a todo here. I wanted to generate a bytearray with my expected final length and just insert bytes where they should go, but I don't think that's a thing you can do with bytearray.

I'll take a look at bytesIO and memory-mapped file. I have no experience with either and I'm curious. But i'll probably leave it for later since it's a performance optimization and I already spent more time than I expected on this